File: agent/java/com/squeeze/agent/net/TrackedHttpsURLConnection.java
package com.squeeze.agent.net;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.security.Permission;
import java.security.Principal;
import java.security.cert.Certificate;
import java.util.List;
import java.util.Map;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLPeerUnverifiedException;
import javax.net.ssl.SSLSocketFactory;

/** Delegating HttpsURLConnection that records the request/response via SqueezeTracker. */
public final class TrackedHttpsURLConnection extends HttpsURLConnection {
    private final HttpsURLConnection d;
    private final SqueezeTracker tracker;
    private boolean requestCaptured = false;

    public TrackedHttpsURLConnection(HttpsURLConnection delegate, SqueezeTracker tracker) {
        super(delegate.getURL());
        this.d = delegate;
        this.tracker = tracker;
    }

    private void captureRequest() {
        if (requestCaptured) return;
        requestCaptured = true;
        try {
            tracker.setMethod(d.getRequestMethod());
            tracker.setRequestHeaders(d.getRequestProperties());
        } catch (Throwable ignored) {}
    }

    private void captureResponse() {
        try { tracker.onResponse(d.getResponseCode(), d.getHeaderFields()); }
        catch (Throwable ignored) {}
    }

    // --- capture points ---
    @Override public void connect() throws IOException { captureRequest(); d.connect(); }
    @Override public OutputStream getOutputStream() throws IOException {
        captureRequest();
        return new TeeStreams.RequestStream(d.getOutputStream(), tracker);
    }
    @Override public int getResponseCode() throws IOException {
        captureRequest();
        int code = d.getResponseCode();
        tracker.onResponse(code, d.getHeaderFields());
        return code;
    }
    @Override public InputStream getInputStream() throws IOException {
        captureRequest();
        try {
            captureResponse();
            return new TeeStreams.ResponseStream(d.getInputStream(), tracker);
        } catch (IOException e) {
            tracker.setError(String.valueOf(e)); tracker.report(); throw e;
        }
    }
    @Override public InputStream getErrorStream() {
        InputStream es = d.getErrorStream();
        return es == null ? null : new TeeStreams.ResponseStream(es, tracker);
    }
    @Override public void disconnect() { tracker.report(); d.disconnect(); }

    // --- HttpURLConnection / URLConnection delegation ---
    @Override public boolean usingProxy() { return d.usingProxy(); }
    @Override public String getResponseMessage() throws IOException { return d.getResponseMessage(); }
    @Override public void setRequestMethod(String m) throws java.net.ProtocolException { d.setRequestMethod(m); }
    @Override public String getRequestMethod() { return d.getRequestMethod(); }
    @Override public void setInstanceFollowRedirects(boolean f) { d.setInstanceFollowRedirects(f); }
    @Override public boolean getInstanceFollowRedirects() { return d.getInstanceFollowRedirects(); }
    @Override public void setFixedLengthStreamingMode(int n) { d.setFixedLengthStreamingMode(n); }
    @Override public void setFixedLengthStreamingMode(long n) { d.setFixedLengthStreamingMode(n); }
    @Override public void setChunkedStreamingMode(int n) { d.setChunkedStreamingMode(n); }
    @Override public Permission getPermission() throws IOException { return d.getPermission(); }
    @Override public String getHeaderFieldKey(int n) { return d.getHeaderFieldKey(n); }
    @Override public String getHeaderField(int n) { return d.getHeaderField(n); }
    @Override public String getHeaderField(String name) { return d.getHeaderField(name); }
    @Override public Map<String, List<String>> getHeaderFields() { return d.getHeaderFields(); }
    @Override public int getHeaderFieldInt(String n, int def) { return d.getHeaderFieldInt(n, def); }
    @Override public long getHeaderFieldDate(String n, long def) { return d.getHeaderFieldDate(n, def); }
    @Override public void setConnectTimeout(int t) { d.setConnectTimeout(t); }
    @Override public int getConnectTimeout() { return d.getConnectTimeout(); }
    @Override public void setReadTimeout(int t) { d.setReadTimeout(t); }
    @Override public int getReadTimeout() { return d.getReadTimeout(); }
    @Override public java.net.URL getURL() { return d.getURL(); }
    @Override public int getContentLength() { return d.getContentLength(); }
    @Override public long getContentLengthLong() { return d.getContentLengthLong(); }
    @Override public String getContentType() { return d.getContentType(); }
    @Override public String getContentEncoding() { return d.getContentEncoding(); }
    @Override public long getExpiration() { return d.getExpiration(); }
    @Override public long getDate() { return d.getDate(); }
    @Override public long getLastModified() { return d.getLastModified(); }
    @Override public Object getContent() throws IOException { return d.getContent(); }
    @Override public Object getContent(Class[] c) throws IOException { return d.getContent(c); }
    @Override public void setDoInput(boolean v) { d.setDoInput(v); }
    @Override public boolean getDoInput() { return d.getDoInput(); }
    @Override public void setDoOutput(boolean v) { d.setDoOutput(v); }
    @Override public boolean getDoOutput() { return d.getDoOutput(); }
    @Override public void setAllowUserInteraction(boolean v) { d.setAllowUserInteraction(v); }
    @Override public boolean getAllowUserInteraction() { return d.getAllowUserInteraction(); }
    @Override public void setUseCaches(boolean v) { d.setUseCaches(v); }
    @Override public boolean getUseCaches() { return d.getUseCaches(); }
    @Override public void setIfModifiedSince(long v) { d.setIfModifiedSince(v); }
    @Override public long getIfModifiedSince() { return d.getIfModifiedSince(); }
    @Override public void setDefaultUseCaches(boolean v) { d.setDefaultUseCaches(v); }
    @Override public boolean getDefaultUseCaches() { return d.getDefaultUseCaches(); }
    @Override public void setRequestProperty(String k, String v) { d.setRequestProperty(k, v); }
    @Override public void addRequestProperty(String k, String v) { d.addRequestProperty(k, v); }
    @Override public String getRequestProperty(String k) { return d.getRequestProperty(k); }
    @Override public Map<String, List<String>> getRequestProperties() { return d.getRequestProperties(); }

    // --- HttpsURLConnection specifics ---
    @Override public String getCipherSuite() { return d.getCipherSuite(); }
    @Override public Certificate[] getLocalCertificates() { return d.getLocalCertificates(); }
    @Override public Certificate[] getServerCertificates() throws SSLPeerUnverifiedException { return d.getServerCertificates(); }
    @Override public Principal getPeerPrincipal() throws SSLPeerUnverifiedException { return d.getPeerPrincipal(); }
    @Override public Principal getLocalPrincipal() { return d.getLocalPrincipal(); }
    @Override public void setHostnameVerifier(HostnameVerifier v) { d.setHostnameVerifier(v); }
    @Override public HostnameVerifier getHostnameVerifier() { return d.getHostnameVerifier(); }
    @Override public void setSSLSocketFactory(SSLSocketFactory f) { d.setSSLSocketFactory(f); }
    @Override public SSLSocketFactory getSSLSocketFactory() { return d.getSSLSocketFactory(); }
}

File: agent/kotlin/com/squeeze/capture/TrackedHttpsURLConnection.kt
package com.squeeze.capture

import java.io.InputStream
import java.io.OutputStream
import java.net.URL
import java.security.Permission
import java.security.Principal
import java.security.cert.Certificate
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLSocketFactory

/** Delegating HttpsURLConnection that records the request/response via SqueezeTracker. */
class TrackedHttpsURLConnection(
    private val d: HttpsURLConnection,
    private val tracker: SqueezeTracker,
) : HttpsURLConnection(d.url) {

    init { tracker.setHttpStack("urlconnection") }


    private var requestCaptured = false

    private fun captureRequest() {
        if (requestCaptured) return
        requestCaptured = true
        try {
            // HttpURLConnection keeps requestMethod == "GET" until connect even when
            // doOutput is set (which implies POST), so correct it here like AOSP does.
            tracker.setMethod(if (d.doOutput && d.requestMethod == "GET") "POST" else d.requestMethod)
            tracker.setRequestHeaders(d.requestProperties)
        } catch (t: Throwable) {}
    }

    // --- capture points ---
    override fun connect() { captureRequest(); d.connect() }
    override fun getOutputStream(): OutputStream { captureRequest(); return RequestStream(d.outputStream, tracker) }
    override fun getResponseCode(): Int {
        captureRequest()
        val code = d.responseCode
        tracker.onResponse(code, d.headerFields)
        return code
    }
    override fun getInputStream(): InputStream {
        captureRequest()
        return try {
            tracker.onResponse(d.responseCode, d.headerFields)
            ResponseStream(d.inputStream, tracker)
        } catch (e: java.io.IOException) {
            tracker.setError(e.toString()); tracker.report(); throw e
        }
    }
    override fun getErrorStream(): InputStream? {
        val es = d.errorStream ?: return null
        return ResponseStream(es, tracker)
    }
    override fun disconnect() { tracker.report(); d.disconnect() }

    // --- HttpURLConnection / URLConnection delegation ---
    override fun usingProxy(): Boolean = d.usingProxy()
    override fun getResponseMessage(): String? = d.responseMessage
    override fun setRequestMethod(method: String?) { d.requestMethod = method }
    override fun getRequestMethod(): String? = d.requestMethod
    override fun setInstanceFollowRedirects(followRedirects: Boolean) { d.instanceFollowRedirects = followRedirects }
    override fun getInstanceFollowRedirects(): Boolean = d.instanceFollowRedirects
    override fun setFixedLengthStreamingMode(contentLength: Int) { d.setFixedLengthStreamingMode(contentLength) }
    override fun setFixedLengthStreamingMode(contentLength: Long) { d.setFixedLengthStreamingMode(contentLength) }
    override fun setChunkedStreamingMode(chunklen: Int) { d.setChunkedStreamingMode(chunklen) }
    override fun getPermission(): Permission? = d.permission
    override fun getHeaderFieldKey(n: Int): String? = d.getHeaderFieldKey(n)
    override fun getHeaderField(n: Int): String? = d.getHeaderField(n)
    override fun getHeaderField(name: String?): String? = d.getHeaderField(name)
    override fun getHeaderFields(): MutableMap<String, MutableList<String>> = d.headerFields
    override fun getHeaderFieldInt(name: String?, default: Int): Int = d.getHeaderFieldInt(name, default)
    override fun getHeaderFieldDate(name: String?, default: Long): Long = d.getHeaderFieldDate(name, default)
    override fun setConnectTimeout(timeout: Int) { d.connectTimeout = timeout }
    override fun getConnectTimeout(): Int = d.connectTimeout
    override fun setReadTimeout(timeout: Int) { d.readTimeout = timeout }
    override fun getReadTimeout(): Int = d.readTimeout
    override fun getURL(): URL = d.url
    override fun getContentLength(): Int = d.contentLength
    override fun getContentLengthLong(): Long = d.contentLengthLong
    override fun getContentType(): String? = d.contentType
    override fun getContentEncoding(): String? = d.contentEncoding
    override fun getExpiration(): Long = d.expiration
    override fun getDate(): Long = d.date
    override fun getLastModified(): Long = d.lastModified
    override fun getContent(): Any? = d.content
    override fun getContent(classes: Array<out Class<*>>?): Any? = d.getContent(classes)
    override fun setDoInput(doInput: Boolean) { d.doInput = doInput }
    override fun getDoInput(): Boolean = d.doInput
    override fun setDoOutput(doOutput: Boolean) { d.doOutput = doOutput }
    override fun getDoOutput(): Boolean = d.doOutput
    override fun setAllowUserInteraction(v: Boolean) { d.allowUserInteraction = v }
    override fun getAllowUserInteraction(): Boolean = d.allowUserInteraction
    override fun setUseCaches(v: Boolean) { d.useCaches = v }
    override fun getUseCaches(): Boolean = d.useCaches
    override fun setIfModifiedSince(v: Long) { d.ifModifiedSince = v }
    override fun getIfModifiedSince(): Long = d.ifModifiedSince
    override fun setDefaultUseCaches(v: Boolean) { d.defaultUseCaches = v }
    override fun getDefaultUseCaches(): Boolean = d.defaultUseCaches
    override fun setRequestProperty(key: String?, value: String?) { d.setRequestProperty(key, value) }
    override fun addRequestProperty(key: String?, value: String?) { d.addRequestProperty(key, value) }
    override fun getRequestProperty(key: String?): String? = d.getRequestProperty(key)
    override fun getRequestProperties(): MutableMap<String, MutableList<String>> = d.requestProperties

    // --- HttpsURLConnection specifics ---
    override fun getCipherSuite(): String? = d.cipherSuite
    override fun getLocalCertificates(): Array<Certificate>? = d.localCertificates
    override fun getServerCertificates(): Array<Certificate> = d.serverCertificates
    override fun getPeerPrincipal(): Principal = d.peerPrincipal
    override fun getLocalPrincipal(): Principal? = d.localPrincipal
    override fun setHostnameVerifier(v: HostnameVerifier?) { d.hostnameVerifier = v }
    override fun getHostnameVerifier(): HostnameVerifier = d.hostnameVerifier
    override fun setSSLSocketFactory(sf: SSLSocketFactory?) { d.sslSocketFactory = sf }
    override fun getSSLSocketFactory(): SSLSocketFactory = d.sslSocketFactory
}
